server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -316,6 +343,13 @@ private boolean invariant() {
         assert relocatingShards.size() == shardRoutingsRelocating.size();
         assert relocatingShards.containsAll(shardRoutingsRelocating);
 
+        final Map<Index, List<ShardRouting>> shardRoutingsByIndex =
+            shards.values().stream().collect(Collectors.groupingBy(ShardRouting::index));
+        assert shardsByIndex.size() == shardRoutingsByIndex.size();
+        assert shardsByIndex.keySet().containsAll(shardRoutingsByIndex.keySet());
+        assert shardsByIndex.entrySet().stream().allMatch(e -> e.getValue().size() == shardRoutingsByIndex.get(e.getKey()).size());
+        assert shardsByIndex.entrySet().stream().allMatch(e -> e.getValue().containsAll(shardRoutingsByIndex.get(e.getKey())));
```

Comment:
Could we make this simpler by:

```suggestion
            shards.values().stream().collect(Collectors.groupingBy(ShardRouting::index, Collectors.toSet()));
        assert shardRoutingsByIndex.equals(shardsByIndex);
```